Example output for https://github.com/kubernetes/kubernetes/pull/126387:

```console
$ hack/apidiff.sh -b /nvme/gopath/src/sigs.k8s.io/controller-runtime 
Checking current working tree for API changes since 22a30e7cbbee31b51da2f5e88833efbfad2c3f6b (= v1.32.0-alpha.0-1134-g22a30e7cbbe).
Installing apidiff into /tmp/tmp.g8XDJbNMbv.
Preparing worktree (detached HEAD 22a30e7cbbe)
Updating files: 100% (26364/26364), done.
HEAD is now at 22a30e7cbbe Merge pull request #127700 from macsko/add_option_waitforpodsprocessed

## .
Incompatible changes:
- ./pkg/controller/resourceclaim.NewController: changed from func(github.com/go-logr/logr.Logger, k8s.io/client-go/kubernetes.Interface, k8s.io/client-go/informers/core/v1.PodInformer, k8s.io/client-go/informers/resource/v1alpha3.PodSchedulingContextInformer, k8s.io/client-go/informers/resource/v1alpha3.ResourceClaimInformer, k8s.io/client-go/informers/resource/v1alpha3.ResourceClaimTemplateInformer) (*Controller, error) to func(context.Context, k8s.io/client-go/kubernetes.Interface, k8s.io/client-go/informers/core/v1.PodInformer, k8s.io/client-go/informers/resource/v1alpha3.PodSchedulingContextInformer, k8s.io/client-go/informers/resource/v1alpha3.ResourceClaimInformer, k8s.io/client-go/informers/resource/v1alpha3.ResourceClaimTemplateInformer) (*Controller, error)
- ./pkg/controller/serviceaccount.NewTokensController: changed from func(k8s.io/client-go/informers/core/v1.ServiceAccountInformer, k8s.io/client-go/informers/core/v1.SecretInformer, k8s.io/client-go/kubernetes.Interface, TokensControllerOptions) (*TokensController, error) to func(context.Context, k8s.io/client-go/informers/core/v1.ServiceAccountInformer, k8s.io/client-go/informers/core/v1.SecretInformer, k8s.io/client-go/kubernetes.Interface, TokensControllerOptions) (*TokensController, error)

## ./staging/src/k8s.io/api
no changes
## ./staging/src/k8s.io/apiextensions-apiserver
Incompatible changes:
- ./pkg/controller/status.NewNamingConditionController: changed from func(k8s.io/apiextensions-apiserver/pkg/client/informers/externalversions/apiextensions/v1.CustomResourceDefinitionInformer, k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset/typed/apiextensions/v1.CustomResourceDefinitionsGetter) *NamingConditionController to func(context.Context, k8s.io/apiextensions-apiserver/pkg/client/informers/externalversions/apiextensions/v1.CustomResourceDefinitionInformer, k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset/typed/apiextensions/v1.CustomResourceDefinitionsGetter) *NamingConditionController

## ./staging/src/k8s.io/apimachinery
no changes
## ./staging/src/k8s.io/apiserver
no changes
## ./staging/src/k8s.io/cli-runtime
no changes
## ./staging/src/k8s.io/client-go
Incompatible changes:
- ./util/workqueue.BucketRateLimiter: changed from TypedBucketRateLimiter[any] to TypedBucketRateLimiter[T comparable]
- ./util/workqueue.DelayingQueueConfig: changed from TypedDelayingQueueConfig[any] to TypedDelayingQueueConfig[T comparable]
- ./util/workqueue.ItemExponentialFailureRateLimiter: changed from TypedItemExponentialFailureRateLimiter[any] to TypedItemExponentialFailureRateLimiter[T comparable]
- ./util/workqueue.ItemFastSlowRateLimiter: changed from TypedItemFastSlowRateLimiter[any] to TypedItemFastSlowRateLimiter[T comparable]
- ./util/workqueue.MaxOfRateLimiter: changed from TypedMaxOfRateLimiter[any] to TypedMaxOfRateLimiter[T comparable]
- ./util/workqueue.QueueConfig: changed from TypedQueueConfig[any] to TypedQueueConfig[T comparable]
- ./util/workqueue.RateLimitingQueueConfig: changed from TypedRateLimitingQueueConfig[any] to TypedRateLimitingQueueConfig[T comparable]
- ./util/workqueue.Type: changed from Typed[any] to Typed[t comparable]
- ./util/workqueue.WithMaxWaitRateLimiter: changed from TypedWithMaxWaitRateLimiter[any] to TypedWithMaxWaitRateLimiter[T comparable]
- ./examples/workqueue.(*Controller).Run: changed from func(int, chan struct{}) to func(context.Context, int)
- ./tools/cache.Controller.RunWithContext: added
- ./tools/cache.DefaultWatchErrorHandler: changed from func(*Reflector, error) to func(context.Context, *Reflector, error)
- ./tools/cache.NewIntegerResourceVersionMutationCache: changed from func(Store, Indexer, time.Duration, bool) MutationCache to func(context.Context, Store, Indexer, time.Duration, bool) MutationCache
- ./tools/cache.SharedInformer.AddEventHandlerWithContext: added
- ./tools/cache.SharedInformer.RunWithContext: added
- ./tools/cache.SharedInformer.SetWatchErrorHandlerWithContext: added
Compatible changes:
- ./tools/cache.(*Reflector).ListAndWatchWithContext: added
- ./tools/cache.(*Reflector).RunWithContext: added
- ./tools/cache.Config.WatchErrorHandlerWithContext: added
- ./tools/cache.DeltaFIFOOptions.Logger: added
- ./tools/cache.HandlerOptions: added
- ./tools/cache.WaitForNamedCacheSyncWithContext: added
- ./tools/cache.WatchErrorHandlerWithContext: added

## ./staging/src/k8s.io/cloud-provider
no changes
## ./staging/src/k8s.io/cluster-bootstrap
no changes
## ./staging/src/k8s.io/code-generator
no changes
## ./staging/src/k8s.io/component-base
no changes
## ./staging/src/k8s.io/component-helpers
no changes
## ./staging/src/k8s.io/controller-manager
no changes
## ./staging/src/k8s.io/cri-api
no changes
## ./staging/src/k8s.io/cri-client
no changes
## ./staging/src/k8s.io/csi-translation-lib
no changes
## ./staging/src/k8s.io/dynamic-resource-allocation
no changes
## ./staging/src/k8s.io/endpointslice
no changes
## ./staging/src/k8s.io/kms
no changes
## ./staging/src/k8s.io/kube-aggregator
no changes
## ./staging/src/k8s.io/kube-controller-manager
no changes
## ./staging/src/k8s.io/kube-proxy
no changes
## ./staging/src/k8s.io/kube-scheduler
no changes
## ./staging/src/k8s.io/kubectl
no changes
## ./staging/src/k8s.io/kubelet
no changes
## ./staging/src/k8s.io/metrics
no changes
## ./staging/src/k8s.io/mount-utils
no changes
## ./staging/src/k8s.io/pod-security-admission
no changes
## ./staging/src/k8s.io/sample-apiserver
no changes
## ./staging/src/k8s.io/sample-cli-plugin
no changes
## ./staging/src/k8s.io/sample-controller
no changes

Some notes about API differences:

Changes in internal packages are usually okay.
However, remember that custom schedulers
and scheduler plugins depend on pkg/scheduler/framework.

API changes in staging repos are more critical.
Try to avoid them as much as possible.
But sometimes changing an API is the lesser evil
and/or the impact on downstream consumers is low.
Use common sense and code searches.

To help with assessing the real-world impact of an
API change, hack/apidiff.sh will now try to builds code in
/nvme/gopath/src/sigs.k8s.io/controller-runtime.

controller-runtime is used because
- It tends to use advanced client-go functionality.
- Breaking it has additional impact on controller
  built on top of it.

This doesn't mean that an API change isn't allowed
if it breaks controller runtime, it just needs additional
scrutiny.

https://github.com/kubernetes-sigs/controller-runtime?tab=readme-ov-file#compatibility
explicitly states that a controller-runtime
release cannot be expected to work with a newer
release of the Kubernetes Go packages.

vvvvvvvvvvvvvvvv /nvme/gopath/src/sigs.k8s.io/controller-runtime vvvvvvvvvvvvvvvvvv
# sigs.k8s.io/controller-runtime/pkg/controller/controllertest
pkg/controller/controllertest/util.go:26:35: cannot use &FakeInformer{} (value of type *FakeInformer) as "k8s.io/client-go/tools/cache".SharedIndexInformer value in variable declaration: *FakeInformer does not implement "k8s.io/client-go/tools/cache".SharedIndexInformer (missing method AddEventHandlerWithContext)
pkg/controller/controllertest/util.go:93:9: cannot use f (variable of type *FakeInformer) as "k8s.io/client-go/tools/cache".SharedIndexInformer value in return statement: *FakeInformer does not implement "k8s.io/client-go/tools/cache".SharedIndexInformer (missing method AddEventHandlerWithContext)

WARNING: Building /nvme/gopath/src/sigs.k8s.io/controller-runtime failed. This may or may not be because of the API changes!
^^^^^^^^^^^^^^^^ /nvme/gopath/src/sigs.k8s.io/controller-runtime ^^^^^^^^^^^^^^^^^^

$ echo $?
1
```



And for https://github.com/kubernetes/kubernetes/pull/127709 (a false positive because of apidiff + generic support):
```console
$ hack/apidiff.sh -b /nvme/gopath/src/sigs.k8s.io/controller-runtime
....
Incompatible changes:
- ./util/workqueue.BucketRateLimiter: changed from TypedBucketRateLimiter[any] to TypedBucketRateLimiter[T comparable]
- ./util/workqueue.DelayingQueueConfig: changed from TypedDelayingQueueConfig[any] to TypedDelayingQueueConfig[T comparable]
- ./util/workqueue.ItemExponentialFailureRateLimiter: changed from TypedItemExponentialFailureRateLimiter[any] to TypedItemExponentialFailureRateLimiter[T comparable]
...

vvvvvvvvvvvvvvvv /nvme/gopath/src/sigs.k8s.io/controller-runtime vvvvvvvvvvvvvvvvvv
/nvme/gopath/src/sigs.k8s.io/controller-runtime builds without errors.
^^^^^^^^^^^^^^^^ /nvme/gopath/src/sigs.k8s.io/controller-runtime ^^^^^^^^^^^^^^^^^^
$ echo $?
1
```

--- a/hack/apidiff.sh
+++ b/hack/apidiff.sh
@@ -26,6 +26,11 @@ Usage: $0 [-r <revision>] [directory ...]"
                   Default is the current working tree instead of a revision.
    -r <revision>: Report change in code added since this revision. Default is
                   the common base of origin/master and HEAD.
+   -b <directory> Build all packages in that directory after replacing
+                  Kubernetes dependencies with the current content of the
+                  staging repo. May be given more than once. Must be an
+                  absolute path.
+                  WARNING: this will modify the go.mod in that directory.
    [directory]:   Check one or more specific directory instead of everything.
 EOF
   exit 1
@@ -40,7 +45,8 @@ source "${KUBE_ROOT}/hack/lib/init.sh"
 
 base=
 target=
-while getopts "r:t:" o; do
+builds=()
+while getopts "r:t:b:" o; do
     case "${o}" in
         r)
             base="${OPTARG}"
@@ -57,6 +63,14 @@ while getopts "r:t:" o; do
                 echo >&2
                 usage
             fi
+            ;;
+       b)
+            if [ ! "${OPTARG}" ]; then
+                echo "ERROR: -${o} needs a non-empty parameter" >&2
+                echo >&2
+                usage
+            fi
+            builds+=("${OPTARG}")
             ;;
         *)
             usage
@@ -136,6 +150,8 @@ output_name () {
 
 # run invokes apidiff once per target and stores the output
 # file(s) in the given directory.
+#
+# shellcheck disable=SC2317 # "Command appears to be unreachable" - gets called indirectly.
 run () {
     out="$1"
     mkdir -p "$out"
@@ -144,31 +160,44 @@ run () {
     done
 }
 
-# runWorktree checks out a specific revision, then invokes run there.
-runWorktree () {
-    local out="$1"
-    local worktree="$2"
-    local rev="$3"
+# inWorktree checks out a specific revision, then invokes the given
+# command there.
+#
+# shellcheck disable=SC2317 # "Command appears to be unreachable" - gets called indirectly.
+inWorktree () {
+    local worktree="$1"
+    shift
+    local rev="$1"
+    shift
 
     # Create a copy of the repo with the specific revision checked out.
-    git worktree add -f -d "${worktree}" "${rev}"
-    # Clean up the copy on exit.
-    kube::util::trap_add "git worktree remove -f ${worktree}" EXIT
+    # Might already have been done before.
+    if ! [ -d "${worktree}" ]; then
+        git worktree add -f -d "${worktree}" "${rev}"
+        # Clean up the copy on exit.
+        kube::util::trap_add "git worktree remove -f ${worktree}" EXIT
+    fi
 
     # Ready for apidiff.
     (
         cd "${worktree}"
-        run "${out}"
+        "$@"
     )
 }
 
+# inTarget runs the given command in the target revision of Kubernetes,
+# checking it out in a work tree if necessary.
+inTarget () {
+    if [ -z "${target}" ]; then
+        "$@"
+    else
+        inWorktree "${KUBE_TEMP}/target" "${target}" "$@"
+    fi
+}
+
 # Dump old and new api state.
-if [ -z "${target}" ]; then
-    run "${KUBE_TEMP}/after"
-else
-    runWorktree "${KUBE_TEMP}/after" "${KUBE_TEMP}/target" "${target}"
-fi
-runWorktree "${KUBE_TEMP}/before" "${KUBE_TEMP}/base" "${base}"
+inTarget run "${KUBE_TEMP}/after"
+inWorktree "${KUBE_TEMP}/base" "${base}" run "${KUBE_TEMP}/before"
 
 # Now produce a report. All changes get reported because exporting some API
 # unnecessarily might also be good to know, but the final exit code will only
@@ -198,5 +227,94 @@ compare () {
 for d in "${targets[@]}"; do
     compare "${d}" "${KUBE_TEMP}/before/$(output_name "${d}")" "${KUBE_TEMP}/after/$(output_name "${d}")"
 done
+
+# tryBuild checks whether some other project builds with the staging repos
+# of the current Kubernetes directory.
+#
+# shellcheck disable=SC2317 # "Command appears to be unreachable" - gets called indirectly.
+tryBuild () {
+    local build="$1"
+
+    # Replace all staging repos, whether the project uses them or not (playing it safe...).
+    local repo
+    for repo in $(cd staging/src; find k8s.io -name go.mod); do
+        local path
+        repo=$(dirname "${repo}")
+        path="$(pwd)/staging/src/${repo}"
+        (
+            cd "$build"
+            go mod edit -replace "${repo}"="${path}"
+        )
+    done
+
+    # We only care about building. Breaking compilation of unit tests is also
+    # annoying, but does not affect downstream consumers.
+    (
+        cd "$build"
+        rm -rf vendor
+        go mod tidy
+        go build ./...
+    )
+}
+
+if [ $res -ne 0 ]; then
+    cat <<EOF
+
+Some notes about API differences:
+
+Changes in internal packages are usually okay.
+However, remember that custom schedulers
+and scheduler plugins depend on pkg/scheduler/framework.
+
+API changes in staging repos are more critical.
+Try to avoid them as much as possible.
+But sometimes changing an API is the lesser evil
+and/or the impact on downstream consumers is low.
+Use common sense and code searches.
+EOF
+
+    if [ ${#builds[@]} -gt 0 ]; then
+
+cat <<EOF
+
+To help with assessing the real-world impact of an
+API change, $0 will now try to build code in
+${builds[@]}.
+EOF
+
+        if [[ "${builds[*]}" =~ controller-runtime ]]; then
+cat <<EOF
+
+controller-runtime is used because
+- It tends to use advanced client-go functionality.
+- Breaking it has additional impact on controller
+  built on top of it.
+
+This doesn't mean that an API change isn't allowed
+if it breaks controller runtime, it just needs additional
+scrutiny.
+
+https://github.com/kubernetes-sigs/controller-runtime?tab=readme-ov-file#compatibility
+explicitly states that a controller-runtime
+release cannot be expected to work with a newer
+release of the Kubernetes Go packages.
+EOF
+        fi
+
+        for build in "${builds[@]}"; do
+            echo
+            echo "vvvvvvvvvvvvvvvv ${build} vvvvvvvvvvvvvvvvvv"
+            if inTarget tryBuild "${build}"; then
+                echo "${build} builds without errors."
+            else
+                cat <<EOF
+
+WARNING: Building ${build} failed. This may or may not be because of the API changes!
+EOF
+            fi
+            echo "^^^^^^^^^^^^^^^^ ${build} ^^^^^^^^^^^^^^^^^^"
+        done
+    fi
+fi
 
 exit "$res"

